csharp/parallel/ParallelExtensionsExtras/TaskSchedulers/QueuedTaskScheduler.cs

```diff
@@ -0,0 +1,590 @@
+        /// The maximum allowed concurrency level of this scheduler.  If custom threads are
+        /// used, this represents the number of created threads.
+        /// </summary>
+        private readonly int _concurrencyLevel;
+        /// <summary>Whether we're processing tasks on the current thread.</summary>
+        private static readonly ThreadLocal<bool> s_taskProcessingThread = new ThreadLocal<bool>();
+
+        // ***
+        // *** For when using a target scheduler
+        // ***
+
+        /// <summary>The scheduler onto which actual work is scheduled.</summary>
+        private readonly TaskScheduler _targetScheduler;
+        /// <summary>The queue of tasks to process when using an underlying target scheduler.</summary>
+        private readonly Queue<Task> _nonthreadsafeTaskQueue;
+        /// <summary>The number of Tasks that have been queued or that are running whiel using an underlying scheduler.</summary>
+        private int _delegatesQueuedOrRunning = 0;
+
+        // ***
+        // *** For when using our own threads
+        // ***
+
+        /// <summary>The threads used by the scheduler to process work.</summary>
+        private readonly Thread[] _threads;
+        /// <summary>The collection of tasks to be executed on our custom threads.</summary>
+        private readonly BlockingCollection<Task> _blockingTaskQueue;
+
+        // ***
+
+        /// <summary>Initializes the scheduler.</summary>
+        public QueuedTaskScheduler() : this(Default, 0) { }
+
+        /// <summary>Initializes the scheduler.</summary>
+        /// <param name="targetScheduler">The target underlying scheduler onto which this sceduler's work is queued.</param>
+        public QueuedTaskScheduler(TaskScheduler targetScheduler) : this(targetScheduler, 0) { }
+
+        /// <summary>Initializes the scheduler.</summary>
+        /// <param name="targetScheduler">The target underlying scheduler onto which this sceduler's work is queued.</param>
+        /// <param name="maxConcurrencyLevel">The maximum degree of concurrency allowed for this scheduler's work.</param>
+        public QueuedTaskScheduler(
+            TaskScheduler targetScheduler,
+            int maxConcurrencyLevel)
+        {
+            if (maxConcurrencyLevel < 0) throw new ArgumentOutOfRangeException(nameof(maxConcurrencyLevel));
+
+            // Initialize only those fields relevant to use an underlying scheduler.  We don't
+            // initialize the fields relevant to using our own custom threads.
+            _targetScheduler = targetScheduler ?? throw new ArgumentNullException("underlyingScheduler");
+            _nonthreadsafeTaskQueue = new Queue<Task>();
+
+            // If 0, use the number of logical processors.  But make sure whatever value we pick
+            // is not greater than the degree of parallelism allowed by the underlying scheduler.
+            _concurrencyLevel = maxConcurrencyLevel != 0 ? maxConcurrencyLevel : Environment.ProcessorCount;
```

Comment:
This could be written as follows, but I don't have strong opinions. Pick which you think is most readable:

```suggestion
            _concurrencyLevel = maxConcurrencyLevel ?? Environment.ProcessorCount;
```

Reply:
I'd have to change the type to `int?` (`Nullable<Int32>`) in order to coalesce, right? I'll leave this as is.